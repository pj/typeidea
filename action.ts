import * as fs from 'fs';

export type FieldTypes = 'string' | 'boolean' | 'number';

export type FieldDefaults = string | boolean | number;

export class Action {
  changeLog: string;
  hash: string | null;

  constructor(changeLog: string, hash: string | null) {
    this.changeLog = changeLog;
    this.hash = hash;
  }

  fieldsToHash(): string {
    return `${this.changeLog}`;
  };

}

// Type Actions
export class NewTypeAction extends Action {
  name: string;
  description: string;

  constructor(
    changeLog: string,
    hash: string | null,
    name: string,
    description: string
  ) {
    super(changeLog, hash);
    this.name = name;
    this.description = description;
  }

  fieldsToHash(): string {
    return `${super.fieldsToHash()}_${this.name}_${this.description}`;
  };

  toString(): string {
    return `NewTypeAction(${this.changeLog}, ${this.hash}, ${this.name}, ${this.description})`;
  }
}

export class RenameFieldTypeAction extends Action {
  typeName: string;
  _from: string;
  to: string;

  constructor(
    changeLog: string,
    hash: string | null,
    typeName: string,
    _from: string,
    to: string,
  ) {
    super(changeLog, hash);
    this.typeName = typeName;
    this._from = _from;
    this.to = to;
  }

  fieldsToHash(): string {
    return `${super.fieldsToHash()}_${this.typeName}_${this._from}_${this.to}`;
  };

  toString(): string {
    return `RenameFieldTypeAction(${this.changeLog}, ${this.hash}, ${this.typeName}, ${this._from}, ${this.to})`;
  }
}

export class RequiredFieldTypeAction extends Action {
  typeName: string;
  name: string;

  constructor(
    changeLog: string,
    hash: string | null,
    typeName: string,
    name: string,
  ) {
    super(changeLog, hash);
    this.typeName = typeName;
    this.name = name;
  }

  fieldsToHash(): string {
    return `${super.fieldsToHash()}_${this.typeName}_${this.name}`;
  };

  toString(): string {
    return `RequiredFieldTypeAction(${this.changeLog}, ${this.hash}, ${this.typeName}, ${this.name})`;
  }
}

export class OptionalFieldTypeAction extends Action {
  typeName: string;
  name: string;

  constructor(
    changeLog: string,
    hash: string | null,
    typeName: string,
    name: string,
  ) {
    super(changeLog, hash);
    this.typeName = typeName;
    this.name = name;
  }

  fieldsToHash(): string {
    return `${super.fieldsToHash()}_${this.typeName}_${this.name}`;
  };

  toString(): string {
    return `OptionalFieldTypeAction(${this.changeLog}, ${this.hash}, ${this.typeName}, ${this.name})`;
  }
}

export class DeleteFieldTypeAction extends Action {
  typeName: string;
  name: string;

  constructor(
    changeLog: string,
    hash: string | null,
    typeName: string,
    name: string,
  ) {
    super(changeLog, hash);
    this.typeName = typeName;
    this.name = name;
  }

  fieldsToHash(): string {
    return `${super.fieldsToHash()}_${this.typeName}_${this.name}`;
  };

  toString(): string {
    return `DeleteFieldTypeAction(${this.changeLog}, ${this.hash}, ${this.typeName}, ${this.name})`;
  }
}

export class SetDefaultFieldTypeAction extends Action {
  typeName: string;
  name: string;
  _default: FieldDefaults;

  constructor(
    changeLog: string,
    hash: string | null,
    name: string,
    typeName: string,
    _default: FieldDefaults,
  ) {
    super(changeLog, hash);
    this.typeName = typeName;
    this.name = name;
    this._default = _default;
  }

  fieldsToHash(): string {
    return `${super.fieldsToHash()}_${this.typeName}_${this.name}_${this._default}`;
  };

  toString(): string {
    return `SetDefaultFieldTypeAction(${this.changeLog}, ${this.hash}, ${this.typeName}, ${this.name}, ${this._default})`;
  }
}

export class RemoveDefaultFieldTypeAction extends Action {
  typeName: string;
  name: string;

  constructor(
    changeLog: string,
    hash: string | null,
    typeName: string,
    name: string,
  ) {
    super(changeLog, hash);
    this.typeName = typeName;
    this.name = name;
  }

  fieldsToHash(): string {
    return `${super.fieldsToHash()}_${this.typeName}_${this.name}`;
  };

  toString(): string {
    return `RemoveDefaultFieldTypeAction(${this.changeLog}, ${this.hash}, ${this.typeName}, ${this.name})`;
  }
}

export class AddFieldTypeAction extends Action {
  typeName: string;
  name: string;
  type: FieldTypes;
  description: string;
  optional: boolean;
  _default: FieldDefaults | null;

  constructor(
    changeLog: string,
    hash: string | null,
    typeName: string,
    name: string,
    type: FieldTypes,
    description: string,
    optional: boolean,
    _default: FieldDefaults | null
  ) {
    super(changeLog, hash);
    this.typeName = typeName;
    this.name = name;
    this.type = type;
    this.description = description;
    this.optional = optional;
    this._default = _default;
  }

  fieldsToHash(): string {
    return `${super.fieldsToHash()}_${this.typeName}_${this.name}_${this.type}_${this.description}_${this.optional}_${this._default}`;
  };

  toString(): string {
    return `AddFieldTypeAction(${this.changeLog}, ${this.hash}, ${this.typeName}, ${this.name}, ${this.type}, ${this.description}, ${this.optional}, ${this._default})`;
  }
}

export class UpdateDescriptionTypeAction extends Action {
  typeName: string;
  name: string;
  description: string;

  constructor(
    changeLog: string,
    hash: string | null,
    typeName: string,
    name: string,
    description: string,
  ) {
    super(changeLog, hash);
    this.typeName = typeName;
    this.name = name;
    this.description = description;
  }

  fieldsToHash(): string {
    return `${super.fieldsToHash()}_${this.typeName}_${this.name}_${this.description}`;
  };

  toString(): string {
    return `UpdateDescriptionTypeAction(${this.changeLog}, ${this.hash}, ${this.typeName}, ${this.name}, ${this.description})`;
  }
}

export class ReferenceFieldTypeAction extends Action {
  typeName: string;
  name: string;
  description: string;
  optional: boolean;
  referenceType: string;
  referenceHash: string;

  constructor(
    changeLog: string,
    hash: string | null,
    typeName: string,
    name: string,
    description: string,
    optional: boolean,
    referenceType: string,
    referenceHash: string,
  ) {
    super(changeLog, hash);
    this.typeName = typeName;
    this.name = name;
    this.description = description;
    this.optional = optional;
    this.referenceType = referenceType;
    this.referenceHash = referenceHash;
  }

  fieldsToHash(): string {
    return `${super.fieldsToHash()}_${this.typeName}_${this.name}_${this.description}_${this.optional}_${this.referenceType}_${this.referenceHash}`;
  };

  toString(): string {
    return `ReferenceFieldTypeAction(${this.changeLog}, ${this.hash}, ${this.typeName}, ${this.name}, ${this.description}, ${this.optional}, ${this.referenceType}, ${this.referenceHash})`;
  }
}

export class GroupAction extends Action {
  typeOrServiceName: string;
  actions: Action[];

  constructor(
    changeLog: string,
    hash: string | null,
    typeOrServiceName: string,
    actions: Action[],
  ) {
    super(changeLog, hash);
    this.typeOrServiceName = typeOrServiceName;
    this.actions = actions;
  }

  fieldsToHash(): string {
    const subHashes: string[] = [this.typeOrServiceName];
    for (const action of this.actions) {
      subHashes.push(action.fieldsToHash());
    }
    return subHashes.join('_');
  };

  toString(): string {
    const formattedActions = this.actions.map(action => action.toString());
    return `GroupAction(${this.changeLog}, ${this.hash}, ${this.typeOrServiceName}, ${formattedActions})`;
  }
}

// Service Definitions
export class NewServiceAction extends Action {
  serviceName: string;
  description: string;
  inputType: string;
  outputType: string;
  inputVersion: string;
  outputVersion: string;

  constructor(
    changeLog: string,
    hash: string | null,
    serviceName: string,
    description: string,
    inputType: string,
    outputType: string,
    inputVersion: string,
    outputVersion: string,
  ) {
    super(changeLog, hash);
    this.serviceName = serviceName;
    this.description = description;
    this.inputType = inputType;
    this.outputType = outputType;
    this.inputVersion = inputVersion;
    this.outputVersion = outputVersion;
  }

  fieldsToHash(): string {
    return `${super.fieldsToHash()}_${this.serviceName}_${this.description}_${this.inputType}_${this.outputType}_${this.inputVersion}_${this.outputVersion}`;
  };

  toString(): string {
    return `NewServiceAction(${this.changeLog}, ${this.hash}, ${this.serviceName}, ${this.description}, ${this.inputType}, ${this.outputType}, ${this.inputVersion}, ${this.outputVersion})`;
  }
}

export class UpdateDescriptionServiceAction extends Action {
  serviceName: string;
  description: string;

  constructor(
    changeLog: string,
    hash: string | null,
    serviceName: string,
    description: string,
  ) {
    super(changeLog, hash);
    this.serviceName = serviceName;
    this.description = description;
  }

  fieldsToHash(): string {
    return `${super.fieldsToHash()}_${this.serviceName}_${this.description}`;
  };

  toString(): string {
    return `UpdateDescriptionServiceAction(${this.changeLog}, ${this.hash}, ${this.serviceName}, ${this.description})`;
  }
}

export class AddInputVersionServiceAction extends Action {
  serviceName: string;
  version: string;

  constructor(
    changeLog: string,
    hash: string | null,
    serviceName: string,
    version: string,
  ) {
    super(changeLog, hash);
    this.serviceName = serviceName;
    this.version = version;
  }

  fieldsToHash(): string {
    return `${super.fieldsToHash()}_${this.serviceName}_${this.version}`;
  };

  toString(): string {
    return `AddInputVersionServiceAction(${this.changeLog}, ${this.hash}, ${this.serviceName}, ${this.version})`;
  }
}

export class RemoveInputVersionServiceAction extends Action {
  serviceName: string;
  version: string;

  constructor(
    changeLog: string,
    hash: string | null,
    serviceName: string,
    version: string,
  ) {
    super(changeLog, hash);
    this.serviceName = serviceName;
    this.version = version;
  }

  fieldsToHash(): string {
    return `${super.fieldsToHash()}_${this.serviceName}_${this.version}`;
  };

  toString(): string {
    return `RemoveInputVersionServiceAction(${this.changeLog}, ${this.hash}, ${this.serviceName}, ${this.version})`;
  }
}

export class DeprecateInputVersionServiceAction extends Action {
  serviceName: string;
  version: string;

  constructor(
    changeLog: string,
    hash: string | null,
    serviceName: string,
    version: string,
  ) {
    super(changeLog, hash);
    this.serviceName = serviceName;
    this.version = version;
  }

  fieldsToHash(): string {
    return `${super.fieldsToHash()}_${this.serviceName}_${this.version}`;
  };

  toString(): string {
    return `DeprecateInputVersionServiceAction(${this.changeLog}, ${this.hash}, ${this.serviceName}, ${this.version})`;
  }
}

export class AddOutputVersionServiceAction extends Action {
  serviceName: string;
  version: string;

  constructor(
    changeLog: string,
    hash: string | null,
    serviceName: string,
    version: string,
  ) {
    super(changeLog, hash);
    this.serviceName = serviceName;
    this.version = version;
  }

  fieldsToHash(): string {
    return `${super.fieldsToHash()}_${this.serviceName}_${this.version}`;
  };

  toString(): string {
    return `AddOutputVersionServiceAction(${this.changeLog}, ${this.hash}, ${this.serviceName}, ${this.version})`;
  }
}

export class RemoveOutputVersionServiceAction extends Action {
  serviceName: string;
  version: string;

  constructor(
    changeLog: string,
    hash: string | null,
    serviceName: string,
    version: string,
  ) {
    super(changeLog, hash);
    this.serviceName = serviceName;
    this.version = version;
  }

  fieldsToHash(): string {
    return `${super.fieldsToHash()}_${this.serviceName}_${this.version}`;
  };

  toString(): string {
    return `RemoveOutputVersionServiceAction(${this.changeLog}, ${this.hash}, ${this.serviceName}, ${this.version})`;
  }
}

export class DeprecateOutputVersionServiceAction extends Action {
  serviceName: string;
  version: string;

  constructor(
    changeLog: string,
    hash: string | null,
    serviceName: string,
    version: string,
  ) {
    super(changeLog, hash);
    this.serviceName = serviceName;
    this.version = version;
  }

  fieldsToHash(): string {
    return `${super.fieldsToHash()}_${this.serviceName}_${this.version}`;
  };

  toString(): string {
    return `DeprecateOutputVersionServiceAction(${this.changeLog}, ${this.hash}, ${this.serviceName}, ${this.version})`;
  }
}

function loadAction(action: any): Action {
  switch(action._action_type) {
    // Services
    case 'NewServiceAction':
      return new NewServiceAction(
        action.changeLog,
        action.hash,
        action.name,
        action.description,
        action.inputType, action.outputType,
        action.inputVersion,
        action.outputVersion,
      );
    case 'UpdateDescriptionServiceAction':
      return new UpdateDescriptionServiceAction(
        action.changeLog,
        action.hash,
        action.serviceName,
        action.description
      );
    case 'AddInputVersionServiceAction':
      return new AddInputVersionServiceAction(
        action.changeLog,
        action.hash,
        action.serviceName,
        action.version
      );
    case 'RemoveInputVersionServiceAction':
      return new RemoveInputVersionServiceAction(
        action.changeLog,
        action.hash,
        action.serviceName,
        action.version
      );
    case 'DeprecateInputVersionServiceAction':
      return new DeprecateInputVersionServiceAction(
        action.changeLog,
        action.hash,
        action.serviceName,
        action.version
      );
    case 'AddOutputVersionServiceAction':
      return new AddOutputVersionServiceAction(
        action.changeLog,
        action.hash,
        action.serviceName,
        action.version
      );
    case 'RemoveOutputVersionServiceAction':
      return new RemoveOutputVersionServiceAction(
        action.changeLog,
        action.hash,
        action.serviceName,
        action.version
      );
    case 'DeprecateOutputVersionServiceAction':
      return new DeprecateOutputVersionServiceAction(
        action.changeLog,
        action.hash,
        action.serviceName,
        action.version
      );
    // Types
    case 'RenameFieldTypeAction':
      return new RenameFieldTypeAction(
        action.changeLog,
        action.hash,
        action.typeName,
        action._from,
        action.to
      );
    case 'RequiredFieldTypeAction':
      return new RequiredFieldTypeAction(
        action.changeLog,
        action.hash,
        action.typeName,
        action.name
      );
    case 'OptionalFieldTypeAction':
      return new OptionalFieldTypeAction(
        action.changeLog,
        action.hash,
        action.typeName,
        action.name
      );
    case 'DeleteFieldTypeAction':
      return new DeleteFieldTypeAction(
        action.changeLog,
        action.hash,
        action.typeName,
        action.name
      );
    case 'SetDefaultFieldTypeAction':
      return new SetDefaultFieldTypeAction(
        action.changeLog,
        action.hash,
        action.typeName,
        action.name,
        action._default
      );
    case 'RemoveDefaultFieldTypeAction':
      return new RemoveDefaultFieldTypeAction(
        action.changeLog,
        action.hash,
        action.typeName,
        action.name
      );
    case 'AddFieldTypeAction':
      return new AddFieldTypeAction(
        action.changeLog,
        action.hash,
        action.typeName,
        action.name,
        action.type,
        action.description,
        action.optional,
        action._default
      );
    case 'UpdateDescriptionTypeAction':
      return new UpdateDescriptionTypeAction(
        action.changeLog,
        action.hash,
        action.typeName,
        action.name,
        action.description
      );
    case 'ReferenceFieldTypeAction':
      return new ReferenceFieldTypeAction(
        action.changeLog,
        action.hash,
        action.typeName,
        action.name,
        action.description,
        action.optional,
        action.referenceType,
        action.referenceHash
      );
    case 'NewTypeAction':
      return new NewTypeAction(
        action.changeLog,
        action.hash,
        action.name,
        action.description
      );
    case 'GroupAction':
      const groupedActions = [];
      for (const subAction of action.actions) {
        groupedActions.push(loadAction(subAction));
      }
      return new GroupAction(
        action.changeLog,
        action.hash,
        action.name,
        groupedActions
      );
    default:
      throw new Error(`Unknown Action: ${action}`)
  }
}

export function loadActionLog(path: string): Array<Action> {
  const actions = require(path);
  const outputActions = [];

  for (const action of actions) {
    const log = loadAction(action);
    outputActions.push(log);
  }

  return outputActions;
}

function loadSimpleAction(fields: any): Action {
  var name, inputType, inputVersion, outputType, outputVersion, changeLog,
    description, hash, serviceName, version, typeName, _from, to, _default,
    type, optional, referenceType, referenceHash, _;
  switch(fields[0]) {
    // Services
    case 'new_service':
      [
        _,
        name,
        inputType,
        inputVersion,
        outputType,
        outputVersion,
        changeLog,
        description,
        hash
      ] = fields;
      return new NewServiceAction(
        changeLog,
        hash,
        name,
        description,
        inputType, outputType,
        inputVersion,
        outputVersion,
      );
    case 'update_service_description':
      [
        _,
        serviceName,
        changeLog,
        description,
        hash
      ] = fields;
      return new UpdateDescriptionServiceAction(
        changeLog,
        hash,
        serviceName,
        description
      );
    case 'add_service_input_version':
      [
        _,
        serviceName,
        version,
        changeLog,
        hash
      ] = fields;
      return new AddInputVersionServiceAction(
        changeLog,
        hash,
        serviceName,
        version
      );
    case 'remove_service_input_version':
      [
        _,
        serviceName,
        version,
        changeLog,
        hash
      ] = fields;
      return new RemoveInputVersionServiceAction(
        changeLog,
        hash,
        serviceName,
        version
      );
    case 'deprecate_service_input_version':
      [
        _,
        serviceName,
        version,
        changeLog,
        hash
      ] = fields;
      return new DeprecateInputVersionServiceAction(
        changeLog,
        hash,
        serviceName,
        version
      );
    case 'add_service_output_version':
      [
        _,
        serviceName,
        version,
        changeLog,
        hash
      ] = fields;
      return new AddOutputVersionServiceAction(
        changeLog,
        hash,
        serviceName,
        version
      );
    case 'remove_service_output_version':
      [
        _,
        serviceName,
        version,
        changeLog,
        hash
      ] = fields;
      return new RemoveOutputVersionServiceAction(
        changeLog,
        hash,
        serviceName,
        version
      );
    case 'deprecate_service_output_version':
      [
        _,
        serviceName,
        version,
        changeLog,
        hash
      ] = fields;
      return new DeprecateOutputVersionServiceAction(
        changeLog,
        hash,
        serviceName,
        version
      );
    // Types
    case 'rename_field':
      [
        _,
        typeName,
        _from,
        to,
        changeLog,
        hash
      ] = fields;
      return new RenameFieldTypeAction(
        changeLog,
        hash,
        typeName,
        _from,
        to
      );
    case 'required_field':
      [
        _,
        typeName,
        name,
        changeLog,
        hash
      ] = fields;
      return new RequiredFieldTypeAction(
        changeLog,
        hash,
        typeName,
        name
      );
    case 'optional_field':
      [
        _,
        typeName,
        name,
        changeLog,
        hash
      ] = fields;
      return new OptionalFieldTypeAction(
        changeLog,
        hash,
        typeName,
        name
      );
    case 'delete_field':
      [
        _,
        typeName,
        name,
        changeLog,
        hash
      ] = fields;
      return new DeleteFieldTypeAction(
        changeLog,
        hash,
        typeName,
        name
      );
    case 'set_field_default':
      [
        _,
        typeName,
        name,
        _default,
        changeLog,
        hash
      ] = fields;
      return new SetDefaultFieldTypeAction(
        changeLog,
        hash,
        typeName,
        name,
        _default
      );
    case 'remove_field_default':
      [
        _,
        typeName,
        name,
        changeLog,
        hash
      ] = fields;
      return new RemoveDefaultFieldTypeAction(
        changeLog,
        hash,
        typeName,
        name
      );
    case 'add_field':
      [
        _,
        typeName,
        name,
        type,
        optional,
        _default,
        description,
        changeLog,
        hash
      ] = fields;
      return new AddFieldTypeAction(
        changeLog,
        hash,
        typeName,
        name,
        type,
        description,
        optional,
        _default
      );
    case 'update_description':
      [
        _,
        typeName,
        name,
        description,
        changeLog,
        hash
      ] = fields;
      return new UpdateDescriptionTypeAction(
        changeLog,
        hash,
        typeName,
        name,
        description
      );
    case 'reference_field':
      [
        _,
        typeName,
        name,
        optional,
        referenceType,
        referenceHash,
        description,
        changeLog,
        hash
      ] = fields;
      return new ReferenceFieldTypeAction(
        changeLog,
        hash,
        typeName,
        name,
        description,
        optional,
        referenceType,
        referenceHash
      );
    case 'new_type':
      [
        _,
        name,
        description,
        changeLog,
        hash
      ] = fields;
      return new NewTypeAction(
        changeLog,
        hash,
        name,
        description
      );
    default:
      throw new Error(`Unknown Action: ${fields}`);
  }
}

export function loadSimpleActionLog(path: string): Array<Action> {
  const contents = fs.readFileSync(path, {'encoding': 'utf8'});
  const lines = contents.split('\n');
  const outputActions = [];
  let currentGroup = null;

  for (const line of lines) {
    if (line === '') {
      continue;
    }
    const action = loadSimpleAction(line.split(','));
    outputActions.push(action);
    //if (line[0] === 'start_group') {
    //  currentGroup = [];
    //} else if (line[0] === 'end_group') {
    //  outputActions.push(new GroupAction(
    //    "todo",
    //    ,
    //    action.state,
    //    currentGroup
    //  ));
    //  currentGroup = null;
    //} else {
    //  if (currentGroup !== null) {
    //    currentGroup.push(action);
    //  } else {
    //    outputActions.push(action);
    //  }
    //}
  }

  return outputActions;
}
