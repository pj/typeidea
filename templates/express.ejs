/**
* **GENERATED CODE DO NOT EDIT!**
*/
import {Request, Response} from "express";
<% services.forEach(function (service) { -%>
/*
* <%= service.name %>
*
* <%= service.description %>
*
* # Change Log:
<% service.changeLog.forEach(function (log, index) { -%>
* <%= index + 1 %>. <%= log %>
<%_ }) -%>
*/
  import {
    <%= service.inputVersions.map(function (inputVersion) {
      return service.inputType + "_V" + inputVersion.version;
    } ).join(',\n') %>
  } from './types';
  import {
    <%= service.outputVersions.map(function (outputVersion) {
      return service.outputType + "_V" + outputVersion.version;
    } ).join(',\n') %>
  } from './types';

  export default function <%= service.name %>(
  app: any,
  func: (
    <%= service.inputVersions.map(function (inputVersion) {
      return "input: " + service.inputType + "_V" + inputVersion.version;
    } ).join(' | ') %>
  )
  =>
  <%=
    service.outputVersions.map(function (outputVersion) {
      return service.outputType + "_V" + outputVersion.version;
    } ).join(' | ');
  %>
  ): void {
    app.get('/<%= service.name %>', (req: Request, res: Response) => {
      const inputMessage = <%= service.inputType %>.deserialize(req.body);
      const response = func(inputMessage);
      const outputMessage = <%= service.outputType %>.serialize(response);
      res.json(outputMessage);
    });
  }
<% }) %>
