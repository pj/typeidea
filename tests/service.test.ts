import * as fs from 'fs';
import * as path from 'path';
import * as request from 'supertest';
import * as express from 'express';

import * as typeidea from '../lib/typeidea';
import * as action from '../lib/action';
import * as generate from '../lib/generate';

const test_dirs = fs.readdirSync('./tests/test_data/services');

for (const dir of test_dirs) {
  const stat = fs.statSync(path.join('./tests', 'test_data', 'services', dir));
  if (!stat.isDirectory()) {
    continue;
  }

  it(dir, async () => {
    fs.mkdirSync(path.join('./runtest', dir), {recursive: true});
    const actionLog = require('../tests/' + path.join('test_data', 'services', dir, 'actions.json'));
    const actions = action.loadActionLogFromList(actionLog);
    const hashes = typeidea.hashActions(actions, false);
    const hashedActions = typeidea.addHashes(actions, hashes, null);

    const [types, services] = generate.generateDefinitions(hashedActions);
    const [generatedTypes, generatedServices] = generate.generateTypescriptBoth(
      types,
      services
    );
    expect(generatedTypes).toMatchSnapshot();
    expect(generatedServices).toMatchSnapshot();
    const typesFile = fs.writeFileSync(
      path.join('runtest', dir, 'types.ts'),
      generatedTypes,
    );
    const servicesFile = fs.writeFileSync(
      path.join('runtest', dir, 'services.ts'),
      generatedServices,
    );

    // import service functions file
    const typesImport = require('../' + path.join('runtest', dir, 'types.ts'));
    const servicesImport = require('../' + path.join('runtest', dir, 'services.ts'));

    const testData = require('../tests/' + path.join('test_data', 'services', dir, 'ServiceImpl.ts'));
    // run express tests
    for (const implementation of testData.implementations) {
      const app = express();
      app.use(express.json());
      implementation[1](app);
      await request(app)
        .post('/' + implementation[0])
        .send({
          'a_field': 'world',
          'version': "TestInputType_V0",
          'hash': "716a611faeb9d2ddaa02e37a5d187183ff4d388f47c740bb2202109b3e3c8fc0"
        })
        .set('Accept', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200)
        .then((response) => expect(response.text).toMatchSnapshot());
    }
  });
}

