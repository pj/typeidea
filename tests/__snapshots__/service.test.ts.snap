// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simple_service 1`] = `
"/**
 * **GENERATED CODE DO NOT EDIT!**
 */
import { Request, Response } from \\"express\\";
/*
 * TestInputType
 * An awesome input type
 * Change Log:
 * 0. Needs this for testing new type
 * 1. Need this for testing add field input
 */
/**

*
* @sealed
*/
class TestInputType_V0 {
  readonly version: number;
  readonly hash: string;

  constructor() {
    this.version = 0;
    this.hash =
      \\"a3ad8a8a6d0df732ef4eece7241c0f253b59ffe694f654f8def25b08d868d74a\\";
  }

  static deserialize(message: any): TestInputType_V0 {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return new TestInputType_V0();
  }

  static serialize(message: TestInputType_V0): string {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return JSON.stringify(message);
  }
}

export {
  TestInputType_V0,
  TestInputType_V0 as TestInputType_Ha3ad8a8a6d0df732ef4eece7241c0f253b59ffe694f654f8def25b08d868d74a
};

/**
 * @param a_field An awesome field on the input type
 *
 * @sealed
 */
class TestInputType_V1 {
  readonly version: number;
  readonly hash: string;
  readonly a_field: string;

  constructor(a_field: string) {
    this.version = 1;
    this.hash =
      \\"716a611faeb9d2ddaa02e37a5d187183ff4d388f47c740bb2202109b3e3c8fc0\\";
    this.a_field = a_field;
  }

  static deserialize(message: any): TestInputType_V1 {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return new TestInputType_V1(message.a_field);
  }

  static serialize(message: TestInputType_V1): string {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return JSON.stringify(message);
  }
}

export {
  TestInputType_V1,
  TestInputType_V1 as TestInputType_H716a611faeb9d2ddaa02e37a5d187183ff4d388f47c740bb2202109b3e3c8fc0
};

export class TestInputType {
  static deserialize(message: any): TestInputType_V0 | TestInputType_V1 {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    switch (message.version) {
      case \\"TestInputType_V0\\":
      case \\"TestInputType_Ha3ad8a8a6d0df732ef4eece7241c0f253b59ffe694f654f8def25b08d868d74a\\":
        return new TestInputType_V0();

      case \\"TestInputType_V1\\":
      case \\"TestInputType_H716a611faeb9d2ddaa02e37a5d187183ff4d388f47c740bb2202109b3e3c8fc0\\":
        return new TestInputType_V1(message.a_field);

      default:
        throw new Error(
          \\"Unknown version error or version not present: \\" + message
        );
    }
  }

  static serialize(message: TestInputType_V0 | TestInputType_V1): string {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return JSON.stringify(message);
  }
}

/*
 * TestOutputType
 * An awesome output type
 * Change Log:
 * 0. Needs this for testing new output type
 * 1. Need this field for testing new output type
 */
/**

*
* @sealed
*/
class TestOutputType_V0 {
  readonly version: number;
  readonly hash: string;

  constructor() {
    this.version = 0;
    this.hash =
      \\"2eeed3cda6513d2557733ee19f52c10ab5114385b733468a9aff6c53ea6bf727\\";
  }

  static deserialize(message: any): TestOutputType_V0 {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return new TestOutputType_V0();
  }

  static serialize(message: TestOutputType_V0): string {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return JSON.stringify(message);
  }
}

export {
  TestOutputType_V0,
  TestOutputType_V0 as TestOutputType_H2eeed3cda6513d2557733ee19f52c10ab5114385b733468a9aff6c53ea6bf727
};

/**
 * @param a_field An awesome field on the output type
 *
 * @sealed
 */
class TestOutputType_V1 {
  readonly version: number;
  readonly hash: string;
  readonly a_field: string;

  constructor(a_field: string) {
    this.version = 1;
    this.hash =
      \\"c503f8a6a87dcd268e82e6cffa2e58db72866839a676faa2df84f470f889ae80\\";
    this.a_field = a_field;
  }

  static deserialize(message: any): TestOutputType_V1 {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return new TestOutputType_V1(message.a_field);
  }

  static serialize(message: TestOutputType_V1): string {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return JSON.stringify(message);
  }
}

export {
  TestOutputType_V1,
  TestOutputType_V1 as TestOutputType_Hc503f8a6a87dcd268e82e6cffa2e58db72866839a676faa2df84f470f889ae80
};

export class TestOutputType {
  static deserialize(message: any): TestOutputType_V0 | TestOutputType_V1 {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    switch (message.version) {
      case \\"TestOutputType_V0\\":
      case \\"TestOutputType_H2eeed3cda6513d2557733ee19f52c10ab5114385b733468a9aff6c53ea6bf727\\":
        return new TestOutputType_V0();

      case \\"TestOutputType_V1\\":
      case \\"TestOutputType_Hc503f8a6a87dcd268e82e6cffa2e58db72866839a676faa2df84f470f889ae80\\":
        return new TestOutputType_V1(message.a_field);

      default:
        throw new Error(
          \\"Unknown version error or version not present: \\" + message
        );
    }
  }

  static serialize(message: TestOutputType_V0 | TestOutputType_V1): string {
    if (message.version === null || message.version === undefined) {
      throw new Error(\\"version not present: \\" + message);
    }
    return JSON.stringify(message);
  }
}
"
`;

exports[`simple_service 2`] = `
"/**
 * **GENERATED CODE DO NOT EDIT!**
 */
import { Request, Response } from \\"express\\";

/*
 * AwesomeService
 *
 * An awesome service
 */
import {
  TestOutputType_V1,
  TestOutputType,
  TestInputType_V1,
  TestInputType
} from \\"./types\\";

const AwesomeServiceDefinitions = new Map();

function AwesomeService(
  func: (input: TestInputType_V1) => TestOutputType_V1
): void {
  AwesomeServiceDefinitions.set(\\"TestInputType_V1\\", func);
}

function AwesomeServiceExpress(app: any): void {
  for (let inputVersion of [\\"TestInputType_V1\\"]) {
    if (!AwesomeServiceDefinitions.has(inputVersion)) {
      throw new Error(
        \\"Service definition required for input version: \\" + inputVersion
      );
    }
  }

  app.post(\\"/AwesomeService\\", (req: Request, res: Response) => {
    const body = req.body;
    switch (body[\\"type\\"] + \\"_\\" + body[\\"version\\"]) {
      case \\"TestInputType_V1\\":
        const inputMessage = TestInputType_V1.deserialize(body);
        const func = AwesomeServiceDefinitions.get(\\"TestInputType_V1\\");
        const response = func(inputMessage);
        const outputMessage = TestOutputType.serialize(response);
        res.json(outputMessage);
        return;
      default:
        throw new Error(\\"Unknown input type: \\" + body);
    }
  });
}

export {
  AwesomeService,
  AwesomeServiceExpress
  // <%= service.name %> as service,
  // serviceName
};
"
`;

exports[`simple_service 3`] = `"\\"{\\\\\\"version\\\\\\":1,\\\\\\"hash\\\\\\":\\\\\\"c503f8a6a87dcd268e82e6cffa2e58db72866839a676faa2df84f470f889ae80\\\\\\",\\\\\\"a_field\\\\\\":\\\\\\"hello world\\\\\\"}\\""`;
